kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkManagerBase.java

```diff
@@ -79,42 +92,67 @@ public SearchResult<T> query(SearchQuery query, Duration queryTimeout) {
               .collect(Collectors.toList());
     }
 
-    List<CompletableFuture<SearchResult<T>>> queries =
+    // Shuffle the chunks to query. The chunkList is ordered, meaning if you had multiple concurrent
+    // queries that need to search the same N chunks, they would all attempt to search the same
+    // chunk at the same time, and then proceed to search the next chunk at the same time.
+    // Randomizing the list of chunks helps reduce contention when attempting to concurrently search
+    // a single IndexSearcher.
+    Collections.shuffle(chunksMatchingQuery);
+
+    List<ListenableFuture<SearchResult<T>>> queries =
         chunksMatchingQuery
             .stream()
             .map(
                 (chunk) ->
-                    CompletableFuture.supplyAsync(
-                            () -> chunk.query(query),
-                            currentTraceContext.executorService(queryExecutorService))
-                        // TODO: this will not cancel lucene query. Use ExitableDirectoryReader
-                        //  in the future and pass this timeout
-                        .orTimeout(queryTimeout.toMillis(), TimeUnit.MILLISECONDS))
-            .map(
-                chunkFuture ->
-                    chunkFuture.exceptionally(
-                        err -> {
-                          LOG.warn("Chunk Query Exception: ", err);
-                          // We catch IllegalArgumentException ( and any other exception that
-                          // represents a parse failure ) and instead of returning an empty result
-                          // we throw back an error to the user
-                          if (err.getCause() instanceof IllegalArgumentException) {
-                            throw (IllegalArgumentException) err.getCause();
-                          }
-                          return errorResult;
-                        }))
-            .collect(Collectors.<CompletableFuture<SearchResult<T>>>toList());
-
-    // TODO: if all fails return error instead of empty and add test
-
-    // Using the spotify library ( this method is much easier to operate then using
-    // CompletableFuture.allOf and converting the CompletableFuture<Void> to
-    // CompletableFuture<List<SearchResult>>
-    CompletableFuture<List<SearchResult<T>>> searchResultFuture =
-        CompletableFutures.allAsList(queries);
+                    queryExecutorService.submit(
+                        currentTraceContext.wrap(
+                            () -> {
+                              try {
+                                if (Thread.interrupted()) {
+                                  LOG.warn(
+                                      "Chunk query thread timed out without starting work, returning error result.");
+                                  return errorResult;
+                                }
```

Comment:
Let's leave this TODO in since it will still be helpful to do.

                        // TODO: this will not cancel lucene query. Use ExitableDirectoryReader
                        //  in the future and pass this timeout

Reply:
This todo is incorrect which is why it was removed. We do indeed now cancel the lucene query by using interrupts - additionally, the `ExitableDirectoryReader` is not appropriate for this use case. 

The `ExitableDirectoryReader` would only be applicable when you want to shutdown the _reader_ after a given timeout https://lucene.apache.org/core/7_4_0/core/org/apache/lucene/index/ExitableDirectoryReader.html. We don't want to shutdown the reader, as we reuse the same reader across multiple queries. We want to timeout the query instead, which is why this comment was removed.